services/services.go

```diff
@@ -0,0 +1,249 @@
+
+		if err != nil {
+
+			return ErrCantWriteFile
+
+		}
+	}
+
+	defer writer.Flush()
+
+	return nil
+
+}
+
+func UserTableDataProvider(db *sqlx.DB, c config.Config) error {
+	users := []database.User{}
+	err := db.Select(&users, "SELECT * FROM users;")
+
+	if err != nil {
+
+		return ErrCantFetchData
+
+	}
+
+	records := make([][]string, len(users)-1)
+
+	for _, record := range users {
+		records = append(records, []string{strconv.Itoa(record.UserID), record.Name})
+	}
+
+	records = append(records, []string{fmt.Sprintf("There are %d records in database", len(users))})
+	err = generateCSV("users", c.FilePath, records[1:])
+
+	if err != nil {
+
+		return err
+
+	}
+
+	return nil
+}
+
+func SalesTableDataProvider(db *sqlx.DB, c config.Config) error {
+	sales := []database.Seller{}
+	err := db.Select(&sales, "SELECT * FROM sales;")
+
+	if err != nil {
+		return ErrCantFetchData
+	}
+
+	records := make([][]string, len(sales)-1)
+
+	for _, record := range sales {
+		records = append(records, []string{strconv.Itoa(record.OrderID), strconv.Itoa(record.UserID), strconv.FormatFloat(record.OrderAmount, 'f', 6, 64)})
+	}
+
+	records = append(records, []string{fmt.Sprintf("There are %d records in database", len(sales))})
+	err = generateCSV("sales", c.FilePath, records[1:])
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func ArchiveFile(source, target string) error {
+	filename := filepath.Base(source)
+	target = filepath.Join(target, fmt.Sprintf("%s.tar", filename))
+	tarfile, err := os.Create(target)
+
+	if err != nil {
+		return err
+	}
+
+	defer tarfile.Close()
+
+	tarball := tar.NewWriter(tarfile)
+
+	defer tarball.Close()
+
+	info, err := os.Stat(source)
+
+	if err != nil {
+		return nil
+	}
+
+	var baseDir string
+
+	if info.IsDir() {
+		baseDir = filepath.Base(source)
+	}
+
+	return filepath.Walk(source,
+		func(path string, info os.FileInfo, err error) error {
+
+			if err != nil {
+
+				return err
+
+			}
+
+			header, err := tar.FileInfoHeader(info, info.Name())
+
+			if err != nil {
+
+				return err
+
+			}
+
+			if baseDir != "" {
+				header.Name = filepath.Join(baseDir, strings.TrimPrefix(path, source))
+			}
+
+			if err := tarball.WriteHeader(header); err != nil {
+
+				return err
+
+			}
+
+			if info.IsDir() {
+
+				return nil
+
+			}
+
+			file, err := os.Open(path)
+
+			if err != nil {
+
+				return err
+
+			}
+
+			defer file.Close()
+
+			_, err = io.Copy(tarball, file)
+
+			return err
+
+		})
+}
+
+func folderExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+
+	if err == nil {
+
+		return true, nil
+
+	}
+
+	if os.IsNotExist(err) {
+
+		return false, nil
+
+	}
+
+	return true, err
+
+}
+
+func ArchiveDatabase(db *sqlx.DB, c config.Config, res chan<- Result, errors chan<- error) {
+	start := time.Now()
+	result := Result{}
+	err := UserTableDataProvider(db, c)
+
+	if err != nil {
+		errors <- fmt.Errorf("Failed to dump user database: %s", err)
+	}
+
+	err = SalesTableDataProvider(db, c)
```

Comment:
`UserTableDataProvider` and `SalesTableDataProvider` - can be concurrent. 
MySql can handle 2 parallel `Select *`
HDD can handle 2 parallel flushes (they probably will not so parallel because one select will faster than another) 
In more complicated case: many parallel flushes can reduce HDD write speed, in this case we can separate flushes from selects and do parallel selects but sequential flushes.